File: apps/web/src/types/parametric.ts
/**
 * Types for the Parametric Insurance Analysis feature.
 */

export type DatasetType = 'ibtracs' | 'hurdat2_atlantic' | 'hurdat2_pacific'

export interface TriggerCriteria {
  min_category?: number
  min_wind_knots?: number
  max_pressure_mb?: number
}

/**
 * Payout structure for parametric insurance.
 * Defines payout amounts based on event intensity.
 */
export interface PayoutStructure {
  basePayout: number  // Base payout amount (USD) or limit for percentage-based
  currency: string    // Currency code (USD, EUR, etc.)
  payoutType: 'binary' | 'percentage' | 'tiered'  // Type of payout calculation
  // Tiered payouts based on event intensity
  tiers: PayoutTier[]
}

export interface PayoutTier {
  id: string
  name: string
  // For hurricanes: category threshold, for earthquakes: magnitude threshold
  minIntensity: number
  maxIntensity?: number
  // For binary or multiplier-based payouts
  payoutMultiplier: number  // Multiplier applied to base payout
  fixedPayout?: number      // Or fixed payout amount
  // For percentage-based payouts (non-binary)
  payoutPercent?: number    // Percentage of base payout (0-100)
}

export interface BoundingBox {
  id: string
  name: string
  north: number  // Max latitude
  south: number  // Min latitude
  east: number   // Max longitude
  west: number   // Min longitude
  color?: string
  trigger?: TriggerCriteria
  payout?: PayoutStructure  // Payout configuration for this zone
}

export interface DatasetInfo {
  id: string
  name: string
  description: string
  basins: string[]
  year_range: [number, number]
  source_url: string
}

export interface HistoricalHurricane {
  storm_id: string
  name: string
  year: number
  basin: string
  max_category: number
  max_wind_knots: number
  min_pressure_mb: number | null
  track: HurricaneTrackPoint[]
  start_date: string
  end_date: string
}

export interface HurricaneTrackPoint {
  timestamp: string
  latitude: number
  longitude: number
  wind_knots: number
  pressure_mb: number | null
  category: number
  status: string
}

export interface BoxIntersection {
  box_id: string
  hurricane: HistoricalHurricane
  entry_point: HurricaneTrackPoint
  exit_point?: HurricaneTrackPoint
  max_intensity_in_box: number  // Max wind speed while in box
  min_pressure_in_box?: number  // Min pressure while in box
  category_at_crossing: number
}

export interface BoxStatistics {
  box_id: string
  box_name: string
  total_hurricanes: number
  qualifying_hurricanes: number  // Hurricanes meeting trigger criteria
  years_analyzed: number
  annual_frequency: number
  qualifying_annual_frequency: number
  category_distribution: Record<number, number>  // Category -> count
  monthly_distribution: Record<number, number>   // Month -> count
  average_intensity_knots: number
  max_intensity_knots: number
  trigger_probability: number  // Probability of at least one qualifying crossing per year
  trigger_criteria?: TriggerCriteria
  dataset: string
}

export interface AnalysisFilters {
  startYear: number
  endYear: number
  minCategory: number
  basin: string | null
  dataset: DatasetType
}

// =====================
// Earthquake Types
// =====================

export type EarthquakeDatasetType = 'usgs_worldwide' | 'usgs_us'

export interface EarthquakeTriggerCriteria {
  min_magnitude?: number
  max_depth_km?: number
  min_depth_km?: number
}

export interface EarthquakeBoundingBox {
  id: string
  name: string
  north: number
  south: number
  east: number
  west: number
  color?: string
  trigger?: EarthquakeTriggerCriteria
  payout?: PayoutStructure  // Payout configuration for this zone
}

export interface EarthquakeDatasetInfo {
  id: string
  name: string
  description: string
  coverage: string
  year_range: [number, number]
  source_url: string
}

export interface HistoricalEarthquake {
  event_id: string
  magnitude: number
  magnitude_type?: string
  place: string
  event_time: string
  latitude: number
  longitude: number
  depth_km: number
  significance: number
  tsunami: number
  url?: string
}

export interface EarthquakeBoxStatistics {
  box_id: string
  box_name: string
  total_earthquakes: number
  qualifying_earthquakes: number
  years_analyzed: number
  annual_frequency: number
  qualifying_annual_frequency: number
  magnitude_distribution: Record<string, number>
  depth_distribution: Record<string, number>
  monthly_distribution: Record<number, number>
  average_magnitude: number
  max_magnitude: number
  average_depth_km: number
  shallowest_depth_km: number
  trigger_probability: number
  trigger_criteria?: EarthquakeTriggerCriteria
  dataset: string
}

export interface EarthquakeAnalysisFilters {
  startYear: number
  endYear: number
  minMagnitude: number
  dataset: EarthquakeDatasetType
}

// =====================
// Stress Testing Types
// =====================

/**
 * Configuration for stress-testing trigger zones.
 * Allows extending boundaries and varying parameters.
 */
export interface StressTestConfig {
  // Boundary extension (percentage or fixed km)
  boundaryExtensionKm: number
  boundaryExtensionPercent: number
  usePerecentageExtension: boolean
  
  // Whether to extend cluster boundaries only (not individual boxes)
  extendClusterOnly: boolean
  
  // Enabled state
  enabled: boolean
}

/**
 * Stress test parameter variations for earthquakes.
 */
export interface EarthquakeStressTestParams {
  // Magnitude variations to test
  magnitudeVariations: number[]  // e.g., [-0.5, 0, +0.5]
  
  // Depth variations (km)
  depthVariations: number[]  // e.g., [-10, 0, +10]
}

/**
 * Stress test parameter variations for hurricanes.
 */
export interface HurricaneStressTestParams {
  // Category variations to test
  categoryVariations: number[]  // e.g., [-1, 0, +1]
  
  // Wind speed variations (knots)
  windVariations: number[]  // e.g., [-10, 0, +10]
}

/**
 * Result from a single stress test scenario.
 */
export interface StressTestResult {
  scenarioName: string
  parameterAdjustments: Record<string, number>
  boundaryExtensionKm: number
  triggerProbability: number
  qualifyingEvents: number
  totalEvents: number
  percentageChange: number  // vs baseline
}

/**
 * Extended box with computed stress test boundaries.
 */
export interface ExtendedBoundingBox {
  original: {
    north: number
    south: number
    east: number
    west: number
  }
  extended: {
    north: number
    south: number
    east: number
    west: number
  }
  extensionKm: number
}

// =====================
// Overall Statistics Types
// =====================

/**
 * Options for calculating overall payout aggregation.
 */
export interface PayoutAggregationOptions {
  // 'worst_only' - each event triggers only the worst (highest payout) box
  // 'capped_100' - sum of payouts per event capped at 100% of limit
  // 'sum_all' - sum all payouts (no cap)
  mode: 'worst_only' | 'capped_100' | 'sum_all'
}

/**
 * Overall statistics aggregated across all boxes.
 */
export interface OverallStatistics {
  totalBoxes: number
  totalEvents: number
  totalQualifyingEvents: number
  yearsAnalyzed: number
  overallAnnualFrequency: number
  overallTriggerProbability: number
  
  // Payout-related aggregations
  expectedAnnualPayout: number
  maxSingleEventPayout: number
  totalHistoricalPayouts: number
  avgPayoutPerEvent: number
  
  // Distribution by event (for deduplication)
  eventsWithMultipleBoxes: number
  avgBoxesPerEvent: number
  
  // Settings used
  aggregationMode: PayoutAggregationOptions['mode']
}

/**
 * Stress test comparison between baseline and extended scenarios.
 */
export interface StressTestComparison {
  baseline: OverallStatistics
  extended: OverallStatistics
  
  // Deltas
  triggerProbabilityDelta: number
  expectedPayoutDelta: number
  qualifyingEventsDelta: number
  
  // Percentage changes
  triggerProbabilityChange: number
  expectedPayoutChange: number
  qualifyingEventsChange: number
}

File: apps/api/app/schemas/parametric.py
"""
Pydantic schemas for parametric insurance analysis.
"""
from datetime import datetime
from typing import Optional, List, Dict, Literal
from pydantic import BaseModel, Field
from enum import Enum


class DatasetType(str, Enum):
    """Available hurricane datasets."""
    IBTRACS = "ibtracs"
    HURDAT2_ATLANTIC = "hurdat2_atlantic"
    HURDAT2_PACIFIC = "hurdat2_pacific"


class TriggerCriteria(BaseModel):
    """Trigger criteria for parametric insurance box."""
    min_category: Optional[int] = Field(None, ge=0, le=5, description="Minimum Saffir-Simpson category (0=TS, 1-5)")
    min_wind_knots: Optional[int] = Field(None, ge=0, description="Minimum wind speed in knots")
    max_pressure_mb: Optional[int] = Field(None, ge=800, le=1100, description="Maximum central pressure in millibars")
    
    def matches(self, category: int, wind_knots: int, pressure_mb: Optional[int]) -> bool:
        """Check if a hurricane point matches the trigger criteria."""
        if self.min_category is not None and category < self.min_category:
            return False
        if self.min_wind_knots is not None and wind_knots < self.min_wind_knots:
            return False
        if self.max_pressure_mb is not None:
            if pressure_mb is None or pressure_mb > self.max_pressure_mb:
                return False
        return True


class BoundingBox(BaseModel):
    """A geographic bounding box for trigger zone analysis."""
    id: str
    name: str
    north: float = Field(..., description="Maximum latitude (north boundary)")
    south: float = Field(..., description="Minimum latitude (south boundary)")
    east: float = Field(..., description="Maximum longitude (east boundary)")
    west: float = Field(..., description="Minimum longitude (west boundary)")
    color: Optional[str] = None
    trigger: Optional[TriggerCriteria] = Field(None, description="Trigger criteria for this box")


class HurricaneTrackPoint(BaseModel):
    """A single point on a hurricane's track."""
    timestamp: datetime
    latitude: float
    longitude: float
    wind_knots: int
    pressure_mb: Optional[int] = None
    category: int
    status: str


class HistoricalHurricane(BaseModel):
    """A historical hurricane with its complete track."""
    storm_id: str
    name: str
    year: int
    basin: str
    max_category: int
    max_wind_knots: int
    min_pressure_mb: Optional[int] = None
    track: List[HurricaneTrackPoint]
    start_date: datetime
    end_date: datetime


class HistoricalHurricaneSummary(BaseModel):
    """Summary version without full track for list views."""
    storm_id: str
    name: str
    year: int
    basin: str
    max_category: int
    max_wind_knots: int
    min_pressure_mb: Optional[int] = None
    start_date: datetime
    end_date: datetime


class BoxIntersection(BaseModel):
    """Information about a hurricane crossing a trigger box."""
    box_id: str
    hurricane: HistoricalHurricaneSummary
    entry_point: HurricaneTrackPoint
    exit_point: Optional[HurricaneTrackPoint] = None
    max_intensity_in_box: int
    category_at_crossing: int


class BoxStatistics(BaseModel):
    """Statistical analysis of hurricanes crossing a trigger box."""
    box_id: str
    box_name: str
    total_hurricanes: int
    qualifying_hurricanes: int = Field(0, description="Hurricanes meeting trigger criteria")
    years_analyzed: int
    annual_frequency: float
    qualifying_annual_frequency: float = Field(0.0, description="Annual frequency of qualifying events")
    category_distribution: Dict[int, int]
    monthly_distribution: Dict[int, int]
    average_intensity_knots: float
    max_intensity_knots: int
    trigger_probability: float = Field(
        ...,
        description="Probability of at least one qualifying hurricane crossing per year"
    )
    trigger_criteria: Optional[TriggerCriteria] = None
    dataset: str = "ibtracs"


class AnalysisRequest(BaseModel):
    """Request parameters for box analysis."""
    box: BoundingBox
    start_year: int = 1980
    end_year: int = 2024
    min_category: int = 0
    basin: Optional[str] = None
    dataset: DatasetType = DatasetType.IBTRACS


class BulkAnalysisRequest(BaseModel):
    """Request parameters for analyzing multiple boxes."""
    boxes: List[BoundingBox]
    start_year: int = 1980
    end_year: int = 2024
    min_category: int = 0
    basin: Optional[str] = None
    dataset: DatasetType = DatasetType.IBTRACS


class YearRange(BaseModel):
    """Available year range for historical data."""
    min_year: int
    max_year: int


class DatasetInfo(BaseModel):
    """Information about an available dataset."""
    id: str
    name: str
    description: str
    basins: List[str]
    year_range: tuple[int, int]
    source_url: str
